File: examples/ios/RunAnywhereAI/RunAnywhereAI/Features/Benchmark/BenchmarkService.swift
//
//  BenchmarkService.swift
//  RunAnywhereAI
//
//  Service for running model benchmarks and collecting metrics
//

import Foundation
import RunAnywhere
import os.log
#if canImport(UIKit)
import UIKit
#endif

// MARK: - Benchmark Service

@MainActor
@Observable
final class BenchmarkService {
    // MARK: - Published State
    
    private(set) var state: BenchmarkState = .idle
    private(set) var progress: BenchmarkProgress?
    private(set) var results: [BenchmarkResult] = []
    private(set) var currentError: Error?
    
    // MARK: - Private State
    
    private let logger = Logger(subsystem: "com.runanywhere.RunAnywhereAI", category: "BenchmarkService")
    private var benchmarkTask: Task<Void, Never>?
    private var startTime: Date?
    private var peakMemoryBytes: Int64 = 0
    
    // MARK: - Public Methods
    
    /// Run LLM benchmarks on the specified models
    func runLLMBenchmark(
        modelIds: [String],
        config: BenchmarkConfig = .default
    ) async throws -> [BenchmarkResult] {
        guard !state.isRunning else {
            throw BenchmarkError.alreadyRunning
        }
        
        state = .preparing
        startTime = Date()
        results = []
        currentError = nil
        
        var allResults: [BenchmarkResult] = []
        
        do {
            for (modelIndex, modelId) in modelIds.enumerated() {
                let result = try await benchmarkSingleModel(
                    modelId: modelId,
                    modelIndex: modelIndex,
                    totalModels: modelIds.count,
                    config: config
                )
                allResults.append(result)
                results = allResults
            }
            
            state = .completed
            
            // Export results to file
            try exportResults(allResults)
            
            logger.info("Benchmark completed: \(allResults.count) models tested")
            return allResults
            
        } catch {
            state = .failed(error: error.localizedDescription)
            currentError = error
            throw error
        }
    }
    
    /// Cancel the current benchmark
    func cancel() {
        benchmarkTask?.cancel()
        state = .idle
        logger.info("Benchmark cancelled")
    }
    
    /// Clear results
    func clearResults() {
        results = []
        state = .idle
        currentError = nil
    }
    
    // MARK: - Private Methods
    
    private func benchmarkSingleModel(
        modelId: String,
        modelIndex: Int,
        totalModels: Int,
        config: BenchmarkConfig
    ) async throws -> BenchmarkResult {
        logger.info("Starting benchmark for model: \(modelId)")
        
        // Get model info
        guard let modelInfo = try await RunAnywhere.availableModels().first(where: { $0.id == modelId }) else {
            throw BenchmarkError.modelNotFound(modelId)
        }
        
        // Load model and measure load time
        let loadStart = CFAbsoluteTimeGetCurrent()
        try await RunAnywhere.loadModel(modelId)
        let loadTimeMs = (CFAbsoluteTimeGetCurrent() - loadStart) * 1000
        
        logger.info("Model loaded in \(loadTimeMs)ms")
        
        // Reset peak memory tracking
        peakMemoryBytes = 0
        
        // Warmup phase
        for i in 0..<config.warmupIterations {
            state = .warmingUp(model: modelInfo.name, iteration: i + 1, total: config.warmupIterations)
            updateProgress(modelIndex: modelIndex, totalModels: totalModels)
            
            _ = try await runSingleInference(
                prompt: config.prompts.first!,
                maxTokens: config.maxTokensList.first!
            )
        }
        
        // Benchmark runs
        var runResults: [SingleRunResult] = []
        let totalRuns = config.prompts.count * config.maxTokensList.count * config.testIterations
        var currentRun = 0
        
        for prompt in config.prompts {
            for maxTokens in config.maxTokensList {
                for iteration in 0..<config.testIterations {
                    currentRun += 1
                    state = .running(
                        model: modelInfo.name,
                        prompt: prompt.id,
                        iteration: currentRun,
                        total: totalRuns
                    )
                    updateProgress(modelIndex: modelIndex, totalModels: totalModels)
                    
                    let result = try await runSingleInference(prompt: prompt, maxTokens: maxTokens)
                    runResults.append(result)
                    
                    // Track peak memory
                    let currentMemory = getMemoryUsage()
                    if currentMemory > peakMemoryBytes {
                        peakMemoryBytes = currentMemory
                    }
                }
            }
        }
        
        // Unload model
        try await RunAnywhere.unloadModel()
        
        // Aggregate results
        let aggregatedResult = aggregateResults(
            runResults: runResults,
            modelInfo: modelInfo,
            loadTimeMs: loadTimeMs,
            config: config
        )
        
        logger.info("Benchmark completed for \(modelId): \(aggregatedResult.avgTokensPerSecond) tok/s")
        
        return aggregatedResult
    }
    
    private func runSingleInference(
        prompt: BenchmarkPrompt,
        maxTokens: Int
    ) async throws -> SingleRunResult {
        let options = LLMGenerationOptions(
            maxTokens: maxTokens,
            temperature: 0.7
        )
        
        let result = try await RunAnywhere.generate(prompt.text, options: options)
        
        return SingleRunResult(
            promptId: prompt.id,
            maxTokens: maxTokens,
            tokensPerSecond: result.tokensPerSecond,
            latencyMs: result.latencyMs,
            ttftMs: result.timeToFirstTokenMs,
            outputTokens: result.tokensUsed,
            inputTokens: result.inputTokens,
            timestamp: Date()
        )
    }
    
    private func aggregateResults(
        runResults: [SingleRunResult],
        modelInfo: ModelInfo,
        loadTimeMs: Double,
        config: BenchmarkConfig
    ) -> BenchmarkResult {
        let tokensPerSecondValues = runResults.map { $0.tokensPerSecond }
        let latencyValues = runResults.map { $0.latencyMs }
        let ttftValues = runResults.compactMap { $0.ttftMs }
        
        // Group by prompt for per-prompt stats
        let promptGroups = Dictionary(grouping: runResults) { $0.promptId }
        let promptResults = promptGroups.map { promptId, results -> PromptAggregatedResult in
            let prompt = config.prompts.first { $0.id == promptId }
            return PromptAggregatedResult(
                promptId: promptId,
                promptCategory: prompt?.category.rawValue ?? "unknown",
                avgTokensPerSecond: results.map { $0.tokensPerSecond }.average(),
                avgLatencyMs: results.map { $0.latencyMs }.average(),
                avgTtftMs: results.compactMap { $0.ttftMs }.average(),
                runCount: results.count
            )
        }
        
        return BenchmarkResult(
            id: UUID().uuidString,
            modelId: modelInfo.id,
            modelName: modelInfo.name,
            framework: modelInfo.framework.rawValue,
            deviceId: getDeviceId(),
            deviceModel: getDeviceModel(),
            osVersion: getOSVersion(),
            sdkVersion: RunAnywhere.version,
            gitCommit: getGitCommit(),
            timestamp: Date(),
            modelLoadTimeMs: loadTimeMs,
            avgTokensPerSecond: tokensPerSecondValues.average(),
            p50TokensPerSecond: tokensPerSecondValues.percentile(50),
            p95TokensPerSecond: tokensPerSecondValues.percentile(95),
            minTokensPerSecond: tokensPerSecondValues.min() ?? 0,
            maxTokensPerSecond: tokensPerSecondValues.max() ?? 0,
            avgTtftMs: ttftValues.average(),
            p50TtftMs: ttftValues.percentile(50),
            p95TtftMs: ttftValues.percentile(95),
            avgLatencyMs: latencyValues.average(),
            p50LatencyMs: latencyValues.percentile(50),
            p95LatencyMs: latencyValues.percentile(95),
            peakMemoryBytes: peakMemoryBytes,
            totalRuns: runResults.count,
            promptResults: promptResults,
            config: config
        )
    }
    
    private func updateProgress(modelIndex: Int, totalModels: Int) {
        guard let startTime = startTime else { return }
        
        let elapsed = Date().timeIntervalSince(startTime)
        let overallProgress: Double
        
        switch state {
        case .warmingUp(_, let iteration, let total):
            let warmupProgress = Double(iteration) / Double(total)
            overallProgress = (Double(modelIndex) + warmupProgress * 0.2) / Double(totalModels)
        case .running(_, _, let iteration, let total):
            let runProgress = Double(iteration) / Double(total)
            overallProgress = (Double(modelIndex) + 0.2 + runProgress * 0.8) / Double(totalModels)
        default:
            overallProgress = Double(modelIndex) / Double(totalModels)
        }
        
        let estimatedRemaining: TimeInterval?
        if overallProgress > 0.1 {
            estimatedRemaining = elapsed / overallProgress * (1 - overallProgress)
        } else {
            estimatedRemaining = nil
        }
        
        progress = BenchmarkProgress(
            state: state,
            overallProgress: overallProgress,
            currentModelIndex: modelIndex,
            totalModels: totalModels,
            elapsedTime: elapsed,
            estimatedRemainingTime: estimatedRemaining
        )
    }
    
    private func exportResults(_ results: [BenchmarkResult]) throws {
        let export = BenchmarkExport(results: results)
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        encoder.dateEncodingStrategy = .iso8601
        
        let data = try encoder.encode(export)
        
        let documentsURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let timestamp = ISO8601DateFormatter().string(from: Date())
            .replacingOccurrences(of: ":", with: "-")
        let filename = "benchmark_\(timestamp).json"
        let fileURL = documentsURL.appendingPathComponent(filename)
        
        try data.write(to: fileURL)
        logger.info("Benchmark results exported to: \(fileURL.path)")
    }
    
    // MARK: - Device Info Helpers
    
    private func getDeviceId() -> String {
        #if os(iOS)
        return UIDevice.current.identifierForVendor?.uuidString ?? "unknown"
        #else
        return ProcessInfo.processInfo.hostName
        #endif
    }
    
    private func getDeviceModel() -> String {
        #if os(iOS)
        var systemInfo = utsname()
        uname(&systemInfo)
        let modelCode = withUnsafePointer(to: &systemInfo.machine) {
            $0.withMemoryRebound(to: CChar.self, capacity: 1) {
                String(cString: $0)
            }
        }
        return modelCode
        #else
        return "Mac"
        #endif
    }
    
    private func getOSVersion() -> String {
        #if os(iOS)
        return UIDevice.current.systemVersion
        #else
        return ProcessInfo.processInfo.operatingSystemVersionString
        #endif
    }
    
    private func getGitCommit() -> String? {
        Bundle.main.infoDictionary?["GitCommit"] as? String
    }
    
    private func getMemoryUsage() -> Int64 {
        var info = mach_task_basic_info()
        var count = mach_msg_type_number_t(MemoryLayout<mach_task_basic_info>.size) / 4
        
        let result = withUnsafeMutablePointer(to: &info) {
            $0.withMemoryRebound(to: integer_t.self, capacity: 1) {
                task_info(mach_task_self_, task_flavor_t(MACH_TASK_BASIC_INFO), $0, &count)
            }
        }
        
        return result == KERN_SUCCESS ? Int64(info.resident_size) : 0
    }
}

// MARK: - Errors

enum BenchmarkError: LocalizedError {
    case alreadyRunning
    case modelNotFound(String)
    case inferenceError(String)
    
    var errorDescription: String? {
        switch self {
        case .alreadyRunning:
            return "A benchmark is already running"
        case .modelNotFound(let id):
            return "Model not found: \(id)"
        case .inferenceError(let message):
            return "Inference error: \(message)"
        }
    }
}

// MARK: - Array Extensions for Statistics

private extension Array where Element == Double {
    func average() -> Double {
        guard !isEmpty else { return 0 }
        return reduce(0, +) / Double(count)
    }
    
    func percentile(_ p: Int) -> Double {
        guard !isEmpty else { return 0 }
        let sorted = self.sorted()
        let index = Int(Double(count - 1) * Double(p) / 100.0)
        return sorted[Swift.min(index, count - 1)]
    }
}

File: examples/android/RunAnywhereAI/app/src/main/java/com/runanywhere/runanywhereai/benchmark/BenchmarkService.kt
package com.runanywhere.runanywhereai.benchmark

import android.content.Context
import android.os.Build
import android.os.Debug
import android.util.Log
import com.runanywhere.sdk.public.RunAnywhere
import com.runanywhere.sdk.public.extensions.LLM.LLMGenerationOptions
import com.runanywhere.sdk.public.extensions.availableModels
import com.runanywhere.sdk.public.extensions.generate
import com.runanywhere.sdk.public.extensions.loadLLMModel
import com.runanywhere.sdk.public.extensions.unloadLLMModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.UUID

/**
 * Service for running model benchmarks and collecting metrics
 */
class BenchmarkService(
    private val context: Context,
) {
    // MARK: - State
    
    private val _state = MutableStateFlow<BenchmarkState>(BenchmarkState.Idle)
    val state: StateFlow<BenchmarkState> = _state.asStateFlow()
    
    private val _progress = MutableStateFlow<BenchmarkProgress?>(null)
    val progress: StateFlow<BenchmarkProgress?> = _progress.asStateFlow()
    
    private val _results = MutableStateFlow<List<BenchmarkResult>>(emptyList())
    val results: StateFlow<List<BenchmarkResult>> = _results.asStateFlow()
    
    private val _error = MutableStateFlow<Throwable?>(null)
    val error: StateFlow<Throwable?> = _error.asStateFlow()
    
    // MARK: - Private State
    
    private var startTimeMs: Long = 0
    private var peakMemoryBytes: Long = 0
    
    private val json = Json {
        prettyPrint = true
        encodeDefaults = true
    }
    
    // MARK: - Public Methods
    
    /**
     * Run LLM benchmarks on the specified models
     */
    suspend fun runLLMBenchmark(
        modelIds: List<String>,
        config: BenchmarkConfig = BenchmarkConfig.DEFAULT,
    ): List<BenchmarkResult> = withContext(Dispatchers.Default) {
        if (_state.value.isRunning) {
            throw BenchmarkException.AlreadyRunning
        }
        
        _state.value = BenchmarkState.Preparing
        startTimeMs = System.currentTimeMillis()
        _results.value = emptyList()
        _error.value = null
        
        val allResults = mutableListOf<BenchmarkResult>()
        
        try {
            modelIds.forEachIndexed { modelIndex, modelId ->
                val result = benchmarkSingleModel(
                    modelId = modelId,
                    modelIndex = modelIndex,
                    totalModels = modelIds.size,
                    config = config
                )
                allResults.add(result)
                _results.value = allResults.toList()
            }
            
            _state.value = BenchmarkState.Completed
            
            // Export results to file
            exportResults(allResults)
            
            Log.i(TAG, "Benchmark completed: ${allResults.size} models tested")
            allResults
            
        } catch (e: Exception) {
            _state.value = BenchmarkState.Failed(e.message ?: "Unknown error")
            _error.value = e
            throw e
        }
    }
    
    /**
     * Cancel the current benchmark
     */
    fun cancel() {
        _state.value = BenchmarkState.Idle
        Log.i(TAG, "Benchmark cancelled")
    }
    
    /**
     * Clear results
     */
    fun clearResults() {
        _results.value = emptyList()
        _state.value = BenchmarkState.Idle
        _error.value = null
    }
    
    // MARK: - Private Methods
    
    private suspend fun benchmarkSingleModel(
        modelId: String,
        modelIndex: Int,
        totalModels: Int,
        config: BenchmarkConfig,
    ): BenchmarkResult {
        Log.i(TAG, "Starting benchmark for model: $modelId")
        
        // Get model info
        val modelInfo = RunAnywhere.availableModels().firstOrNull { it.id == modelId }
            ?: throw BenchmarkException.ModelNotFound(modelId)
        
        // Load model and measure load time
        val loadStart = System.nanoTime()
        RunAnywhere.loadLLMModel(modelId)
        val loadTimeMs = (System.nanoTime() - loadStart) / 1_000_000.0
        
        Log.i(TAG, "Model loaded in ${loadTimeMs}ms")
        
        // Reset peak memory tracking
        peakMemoryBytes = 0
        
        // Warmup phase
        for (i in 0 until config.warmupIterations) {
            _state.value = BenchmarkState.WarmingUp(
                model = modelInfo.name,
                iteration = i + 1,
                total = config.warmupIterations
            )
            updateProgress(modelIndex, totalModels)
            
            runSingleInference(
                prompt = config.prompts.first(),
                maxTokens = config.maxTokensList.first()
            )
        }
        
        // Benchmark runs
        val runResults = mutableListOf<SingleRunResult>()
        val totalRuns = config.prompts.size * config.maxTokensList.size * config.testIterations
        var currentRun = 0
        
        for (prompt in config.prompts) {
            for (maxTokens in config.maxTokensList) {
                repeat(config.testIterations) {
                    currentRun++
                    _state.value = BenchmarkState.Running(
                        model = modelInfo.name,
                        prompt = prompt.id,
                        iteration = currentRun,
                        total = totalRuns
                    )
                    updateProgress(modelIndex, totalModels)
                    
                    val result = runSingleInference(prompt, maxTokens)
                    runResults.add(result)
                    
                    // Track peak memory
                    val currentMemory = getMemoryUsage()
                    if (currentMemory > peakMemoryBytes) {
                        peakMemoryBytes = currentMemory
                    }
                }
            }
        }
        
        // Unload model
        RunAnywhere.unloadLLMModel()
        
        // Aggregate results
        val aggregatedResult = aggregateResults(
            runResults = runResults,
            modelInfo = modelInfo,
            loadTimeMs = loadTimeMs,
            config = config
        )
        
        Log.i(TAG, "Benchmark completed for $modelId: ${aggregatedResult.avgTokensPerSecond} tok/s")
        
        return aggregatedResult
    }
    
    private suspend fun runSingleInference(
        prompt: BenchmarkPrompt,
        maxTokens: Int,
    ): SingleRunResult {
        val options = LLMGenerationOptions(
            maxTokens = maxTokens,
            temperature = 0.7f
        )
        
        val result = RunAnywhere.generate(prompt.text, options)
        
        return SingleRunResult(
            promptId = prompt.id,
            maxTokens = maxTokens,
            tokensPerSecond = result.tokensPerSecond,
            latencyMs = result.latencyMs,
            ttftMs = result.timeToFirstTokenMs,
            outputTokens = result.tokensUsed,
            inputTokens = result.inputTokens,
            timestamp = System.currentTimeMillis()
        )
    }
    
    private fun aggregateResults(
        runResults: List<SingleRunResult>,
        modelInfo: com.runanywhere.sdk.public.extensions.Models.ModelInfo,
        loadTimeMs: Double,
        config: BenchmarkConfig,
    ): BenchmarkResult {
        val tokensPerSecondValues = runResults.map { it.tokensPerSecond }
        val latencyValues = runResults.map { it.latencyMs }
        val ttftValues = runResults.mapNotNull { it.ttftMs }
        
        // Group by prompt for per-prompt stats
        val promptGroups = runResults.groupBy { it.promptId }
        val promptResults = promptGroups.map { (promptId, results) ->
            val prompt = config.prompts.firstOrNull { it.id == promptId }
            PromptAggregatedResult(
                promptId = promptId,
                promptCategory = prompt?.category?.name ?: "unknown",
                avgTokensPerSecond = results.map { it.tokensPerSecond }.average(),
                avgLatencyMs = results.map { it.latencyMs }.average(),
                avgTtftMs = results.mapNotNull { it.ttftMs }.averageOrZero(),
                runCount = results.size
            )
        }
        
        return BenchmarkResult(
            id = UUID.randomUUID().toString(),
            modelId = modelInfo.id,
            modelName = modelInfo.name,
            framework = modelInfo.framework.name,
            deviceId = getDeviceId(),
            deviceModel = getDeviceModel(),
            osVersion = getOSVersion(),
            sdkVersion = RunAnywhere.version,
            gitCommit = getGitCommit(),
            timestamp = System.currentTimeMillis(),
            modelLoadTimeMs = loadTimeMs,
            avgTokensPerSecond = tokensPerSecondValues.average(),
            p50TokensPerSecond = tokensPerSecondValues.percentile(50),
            p95TokensPerSecond = tokensPerSecondValues.percentile(95),
            minTokensPerSecond = tokensPerSecondValues.minOrNull() ?: 0.0,
            maxTokensPerSecond = tokensPerSecondValues.maxOrNull() ?: 0.0,
            avgTtftMs = ttftValues.averageOrZero(),
            p50TtftMs = ttftValues.percentile(50),
            p95TtftMs = ttftValues.percentile(95),
            avgLatencyMs = latencyValues.average(),
            p50LatencyMs = latencyValues.percentile(50),
            p95LatencyMs = latencyValues.percentile(95),
            peakMemoryBytes = peakMemoryBytes,
            totalRuns = runResults.size,
            promptResults = promptResults,
            config = config
        )
    }
    
    private fun updateProgress(modelIndex: Int, totalModels: Int) {
        val elapsed = System.currentTimeMillis() - startTimeMs
        val overallProgress: Float = when (val currentState = _state.value) {
            is BenchmarkState.WarmingUp -> {
                val warmupProgress = currentState.iteration.toFloat() / currentState.total
                (modelIndex + warmupProgress * 0.2f) / totalModels
            }
            is BenchmarkState.Running -> {
                val runProgress = currentState.iteration.toFloat() / currentState.total
                (modelIndex + 0.2f + runProgress * 0.8f) / totalModels
            }
            else -> modelIndex.toFloat() / totalModels
        }
        
        val estimatedRemaining: Long? = if (overallProgress > 0.1f) {
            ((elapsed / overallProgress) * (1 - overallProgress)).toLong()
        } else {
            null
        }
        
        _progress.value = BenchmarkProgress(
            state = _state.value,
            overallProgress = overallProgress,
            currentModelIndex = modelIndex,
            totalModels = totalModels,
            elapsedTimeMs = elapsed,
            estimatedRemainingTimeMs = estimatedRemaining
        )
    }
    
    private fun exportResults(results: List<BenchmarkResult>) {
        try {
            val export = BenchmarkExport(results = results)
            val jsonString = json.encodeToString(export)
            
            val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH-mm-ss", Locale.US)
            val timestamp = dateFormat.format(Date())
            val filename = "benchmark_$timestamp.json"
            
            val file = File(context.getExternalFilesDir(null), filename)
            file.writeText(jsonString)
            
            Log.i(TAG, "Benchmark results exported to: ${file.absolutePath}")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to export results", e)
        }
    }
    
    // MARK: - Device Info Helpers
    
    private fun getDeviceId(): String {
        return android.provider.Settings.Secure.getString(
            context.contentResolver,
            android.provider.Settings.Secure.ANDROID_ID
        ) ?: "unknown"
    }
    
    private fun getDeviceModel(): String {
        return "${Build.MANUFACTURER} ${Build.MODEL}"
    }
    
    private fun getOSVersion(): String {
        return "Android ${Build.VERSION.RELEASE} (API ${Build.VERSION.SDK_INT})"
    }
    
    private fun getGitCommit(): String? {
        return try {
            context.packageManager.getApplicationInfo(
                context.packageName,
                android.content.pm.PackageManager.GET_META_DATA
            ).metaData?.getString("GIT_COMMIT")
        } catch (e: Exception) {
            null
        }
    }
    
    private fun getMemoryUsage(): Long {
        return Debug.getNativeHeapAllocatedSize()
    }
    
    companion object {
        private const val TAG = "BenchmarkService"
    }
}

// MARK: - Exceptions

sealed class BenchmarkException(message: String) : Exception(message) {
    data object AlreadyRunning : BenchmarkException("A benchmark is already running")
    data class ModelNotFound(val modelId: String) : BenchmarkException("Model not found: $modelId")
    data class InferenceError(val details: String) : BenchmarkException("Inference error: $details")
}

// MARK: - List Extensions for Statistics

private fun List<Double>.averageOrZero(): Double {
    return if (isEmpty()) 0.0 else average()
}

private fun List<Double>.percentile(p: Int): Double {
    if (isEmpty()) return 0.0
    val sorted = sorted()
    val index = ((size - 1) * p / 100.0).toInt()
    return sorted[minOf(index, size - 1)]
}
